pending-documentation.md

```diff
@@ -1157,45 +1198,31 @@ rule X +Int (Y +Int Z) => (X +Int Z) +Int Y
    andBool #isVariable(Y)
 ```
 
-### `simplification` attribute (Haskell backend)
-
-The simplification attribute identifies axioms that are useful for simplifying
-configurations, without being part of the main semantics. When a rule is tagged
-as `simplification`, the Haskell backend will only apply that rule if:
+### `concrete` and `symbolic` attributes (Haskell backend)
 
--   the rule lhs _matches_ the subterm of interest, and
--   the side condition has no remainder given the current top-level predicate
-    (that is, the current top-level predicate _implies_ the side condition of
-    the rule).
+Sometimes you only want a rule to apply if some or all arguments are concrete
+(not symbolic). This is done with the `concrete` attribute. Conversely, the
+`symbolic` attribute will allow a rule to apply only when some arguments are not
+concrete. These attributes should only be given with the `simplification`
+attribute.
 
-Note that the `simplification` attribute can be applied to _any_ rule, not just
-function rules, and that `simplification` rules are tried _before_ rules from
-the semantic definition.
-
-For example, for the following definition:
+For example, the following will only re-associate terms when all arguments
+are concrete:
 
 ```k
-    syntax WordStack ::= Int ":" WordStack | ".WordStack"
-    syntax Int ::= sizeWordStack    ( WordStack       ) [function]
-                 | sizeWordStackAux ( WordStack , Int ) [function]
- // --------------------------------------------------------------
-    rule sizeWordStack(WS) => sizeWordStackAux(WS, 0)
-
-    rule sizeWordStackAux(.WordStack, N) => N
-    rule sizeWordStackAux(W : WS    , N) => sizeWordStackAux(WS, N +Int 1)
+rule X +Int (Y +Int Z) => (X +Int Y) +Int Z [simplification, concrete]
 ```
 
-We might add the following simplification lemma:
+And the following rules will only re-associate terms when it will end up
+grouping concrete sub-terms:
 
 ```k
-    rule sizeWordStackAux(WS, N) => N +Int sizeWordStackAux(WS, 0)
-      requires N =/=Int 0
-      [simplification]
-```
+rule X +Int (Y +Int Z) => (X +Int Y) +Int Z
+  [simplification, concrete(X), concrete(Y), symbolic(Z)]
```

Comment:
I thought we supported something like `concrete(X, Y), symbolic(Z)` instead of having them all be separate?

Reply:
Yes, looking at `domains.k` we do support grouping them. Let's use the grouped version, to encourage better styling for people.